File: leetcode3/이진희/498. Diagonal Traverse.java
/* 

1. 아이디어 : 같은 대각선은 r+c가 같다는 특징이 있으므로 대각선 번호 k를 0부터 m + n - 2까지 순회한다
              if문을 사용하여 어떤 방향으로 순회할지 결정한다
              k가 짝수이면 아래에서 위로, k가 홀수이면 위에서 아래로 이동한다

2. 시간복잡도 : O(M*N)

3. 자료구조/알고리즘 : 완전탐색

 */

class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;

        int[] ans = new int[m*n];
        int idx = 0;

        for (int k=0; k < m+n-1; k++) {
            if (k%2 == 0) {
                for (int r=m-1; r>=0; r--) {
                    int c = k-r;
                    if (c>=0 && c<n) {
                        ans[idx++] = mat[r][c];
                    }
                }
            } else {
                for (int r=0; r<m; r++) {
                    int c = k-r;
                    if (c>=0 && c<n) {
                        ans[idx++] = mat[r][c];
                    }
                }
            }
        }
        return ans;
    }
}

File: leetcode3/최원준/498. Diagonal Traverse.py
#

'''
1. 아이디어 :
-

2. 시간복잡도 :
    O(n * m)

3. 자료구조/알고리즘 :
-

'''

class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        ans = []
        n = len(mat)
        m = len(mat[0])

        def get_next(x: int, y: int, going_up: bool):
            if going_up:
                if x == 0:
                    if y == m - 1:
                        return [x + 1, y]
                    else:
                        return [x, y + 1]
                elif y == m - 1:
                    return [x + 1, y]
                else:
                    return [x - 1, y + 1]

            else: 
                if y == 0:
                    if x == n - 1:
                        return [x, y + 1]
                    else:
                        return [x + 1, y]
                elif x == n - 1:
                    return [x, y + 1]
                else:
                    return [x + 1, y - 1]

        x = y = 0
        going_up = True

        for _ in range(n * m):
            ans.append(mat[x][y])
            nx, ny = get_next(x, y, going_up)

            if nx != x - 1 and nx != x + 1:
                going_up = not going_up
            elif ny != y - 1 and ny != y + 1:
                going_up = not going_up

            x, y = nx, ny

        return ans
